### PR TITLE
fix: some bag fix

### PR DESCRIPTION
**Pull request by Romaleso261981**

### CODE REVIEW

File: src/app/App.tsx [hunk 1]
```diff
@@ -15,28 +20,50 @@ const AuthPage = lazy(() => import('../pages/Auth/Auth'));
 
 const App: React.FC = () => {
   const [showAdminBoard, setShowAdminBoard] = useState<boolean>(false);
-  const [user, setUser] = useState<User | null>(null);
+  const [dataUsers, setDataUsers] = useState<UserData[]>([]);
+  const [currentUsers, setCurrentUsers] = useState<User | null>(null);
+
+  const navigate = useNavigate();
+  const setIsRegistered = useSelector(useSelectSetIsRegistered);
```
Comment (suveiko) [inline src/app/App.tsx]: create useAppSelector instead of useSelector

image

File: src/app/App.tsx [hunk 2]
```diff
@@ -15,28 +20,50 @@ const AuthPage = lazy(() => import('../pages/Auth/Auth'));
...
+  const getData = async () => {
+    const data = await getFirestoreData<UserData>(DatabasePaths.USERS, 20);
```
Comment (suveiko) [inline src/app/App.tsx]: move this logic to redux

File: src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import { getCurrentUser } from '../../../../../../shared/helpers/getCurrentUser';
+import { logOut } from '../../../../../auth/authSlice';
+
+import classes from './UserInfoIcons.module.css';
+
+type Props = {
+  toggleUserShow: () => void;
+};
+
+export const UserInfoIcons = (props: Props) => {
```
Comment (suveiko) [inline src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx]:
```suggestion
export const UserInfoIcons = ({ toggleUserShow }: Props) => {
```

File: src/features/user/types.ts [hunk 4]
```diff
@@ -32,6 +34,8 @@ export type AuthError = {
 
 export type UserState = {
   user: User | null | undefined;
+  users: [] | UserData[];
```
Comment (suveiko) [inline src/features/user/types.ts]:
```suggestion
  users: UserData[];
```

File: src/features/user/userSlice.ts [hunk 5]
```diff
@@ -1,13 +1,55 @@
...
+export const editUser = createAsyncThunk(
+  'user/editUser',
+  async ({ id, editedUser }: Partial<ResDataUser>, { rejectWithValue }) => {
+    try {
+      if (!id) {
+        throw new Error('Відсутній ідентифікатор користувача.');
+      }
+      const docRef = doc(db, 'users', id);
+
+      console.log({ ...editedUser });
```
Comment (suveiko) [inline src/features/user/userSlice.ts]: remove log

File: src/pages/Admin/Admin.tsx [hunk 6]
```diff
@@ -1,30 +1,19 @@
...
-    }
-  };
+type Props = {
+  dataUsers: UserData[];
+};
 
-  useEffect(() => {
-    getData();
-  }, []);
+const Admin = (props: Props) => {
```
Comment (suveiko) [inline src/pages/Admin/Admin.tsx]:
```suggestion
const Admin = ({ dataUsers }: Props) => {
```

File: src/pages/Dashboard/ui/components/TableSelection/TableSelection.tsx [hunk 7]
```diff
@@ -15,6 +28,7 @@ type Props = {
 export const TableSelection = (props: Props) => {
   const { userData = [] } = props;
   const navigate = useNavigate();
+  const dispach: AppDispatch = useDispatch();
```
Comment (suveiko) [inline src/pages/Dashboard/ui/components/TableSelection/TableSelection.tsx]:
```suggestion
  const dispatch = useAppDispatch();
```

File: src/pages/Dashboard/ui/components/TableSelection/TableSelection.tsx [hunk 8]
```diff
@@ -70,6 +88,20 @@ export const TableSelection = (props: Props) => {
           <LoadingOverlay />
           <Table.Td>{item.balans}</Table.Td>
         </Group>
+        <Group gap={0} justify="flex-end">
+          <ActionIcon variant="subtle" color="gray">
+            <IconPencil style={{ width: rem(16), height: rem(16) }} stroke={1.5} />
+          </ActionIcon>
+          <ActionIcon variant="subtle" color="red">
+            <IconTrash
+              onClick={() => {
```
Comment (suveiko) [inline src/pages/Dashboard/ui/components/TableSelection/TableSelection.tsx]: create function for this

File: src/pages/UserDetail/UserDetail.tsx [hunk 9]
```diff
@@ -1,34 +1,40 @@
...
-        }
-      })
-      // eslint-disable-next-line no-console
-      .catch((err) => console.log(`error ${err}`));
+  const { id = '' } = useParams<UserParams>();
+  const [user, setUser] = useState<UserData | null>(null);
+  const [isEditing, setIsEditing] = useState(false);
+  const [editedUser, setEditedUser] = useState<UserData | null>(null);
+
+  const dispach: AppDispatch = useDispatch();
```
Comment (suveiko) [inline src/pages/UserDetail/UserDetail.tsx]: move dispatch to the top of the component and fix naming dispach -> dispatch

File: src/pages/UserDetail/UserDetail.tsx [hunk 10]
```diff
@@ -37,26 +43,141 @@ export const UserDetail: React.FC = () => {
...
+
+      {isEditing ? (
+        <>
+          <div>
+            <label htmlFor="editedNickName">nickName:</label>
+            <input
+              type="text"
+              id="editedNickName"
+              value={editedUser?.nickName || ''}
+              onChange={(e) => handleInputChange('nickName', e.target.value)}
```
Comment (suveiko) [inline src/pages/UserDetail/UserDetail.tsx]: move outside

File: src/pages/UserDetail/UserDetail.tsx [hunk 11]
```diff
@@ -37,26 +43,141 @@ export const UserDetail: React.FC = () => {
...
+            />
+          </div>
+          <div>
+            <label htmlFor="editedRule">Rule:</label>
+            <input
+              type="text"
+              id="editedRule"
+              value={editedUser?.rule || ''}
+              onChange={(e) => handleInputChange('rule', e.target.value)}
```
Comment (suveiko) [inline src/pages/UserDetail/UserDetail.tsx]: move outside

File: src/pages/UserDetail/UserDetail.tsx [hunk 12]
```diff
@@ -37,26 +43,141 @@ export const UserDetail: React.FC = () => {
...
+            />
+          </div>
+          <div>
+            <label htmlFor="editedBalans">Balans:</label>
+            <input
+              type="text"
+              id="editedBalans"
+              value={editedUser?.balans || ''}
+              onChange={(e) => handleInputChange('balans', e.target.value)}
```
Comment (suveiko) [inline src/pages/UserDetail/UserDetail.tsx]: move outside

File: src/shared/helpers/Redirect.ts [hunk 13]
```diff
@@ -0,0 +1,8 @@
+import { useNavigate } from 'react-router-dom';
+
+export const Redirect = (path: string) => {
```
Comment (suveiko) [inline src/shared/helpers/Redirect.ts]: why you create this function?

File: src/shared/helpers/editUser.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { doc, updateDoc } from 'firebase/firestore';
+
+import { db } from '../../integations/firebase';
+import { UserData } from '../types/Types';
+
+export const editUser = async (userId: string, editedUser: UserData) => {
+  try {
+    console.log(userId, editedUser);
```
Comment (suveiko) [inline src/shared/helpers/editUser.ts]: remove log

File: src/app/App.tsx [hunk 15]
```diff
@@ -2,43 +2,65 @@ import { Suspense, lazy, useEffect, useState } from 'react';
...
 
 const App: React.FC = () => {
+  // const navigate = useNavigate();
+  const dispach = useAppDispatch();
   const [showAdminBoard, setShowAdminBoard] = useState<boolean>(false);
-  const [user, setUser] = useState<User | null>(null);
 
-  useEffect(() => {
-    const user = getUser();
+  const user = useAppSelector(useSelectUser) ?? null;
```
Comment (remadyt) [inline src/app/App.tsx]: move null to useSelectUser

File: src/app/App.tsx [hunk 16]
```diff
@@ -2,43 +2,65 @@ import { Suspense, lazy, useEffect, useState } from 'react';
...
+import { AuthChecker } from './AuthChecker';
+import { useAppDispatch, useAppSelector } from './store';
 
 const Main = lazy(() => import('../pages/Main/Main'));
 const NotFound = lazy(() => import('../pages/NoFound/NoFound'));
 const Admin = lazy(() => import('../pages/Admin/Admin'));
 const AuthPage = lazy(() => import('../pages/Auth/Auth'));
 
 const App: React.FC = () => {
+  // const navigate = useNavigate();
```
Comment (remadyt) [inline src/app/App.tsx]: why need navigate?

File: src/app/App.tsx [hunk 17]
```diff
@@ -2,43 +2,65 @@ import { Suspense, lazy, useEffect, useState } from 'react';
...
   return (
     <Suspense fallback={<Spiner />}>
       <Routes>
         <Route path="/" element={<RootLayout />}>
-          <Route index element={user ? <Main /> : <AuthPage />} />
-          <Route path="auth" element={<AuthPage />} />
+          <Route
+            index
+            element={
+              <AuthChecker path="/auth" comp>
```
Comment (remadyt) [inline src/app/App.tsx]: use enum for Paths

File: src/app/App.tsx [hunk 18]
```diff
@@ -2,43 +2,65 @@ import { Suspense, lazy, useEffect, useState } from 'react';
...
+            }
+          />
           <Route
             path="admin"
-            element={showAdminBoard ? <Admin /> : <h1>Ви немаєте права доступу</h1>}
+            element={
+              showAdminBoard ? (
+                <Admin dataUsers={dataUsers} />
+              ) : (
+                <h1>Ви немаєте права доступу</h1>
```
Comment (remadyt) [inline src/app/App.tsx]: transaltes?

File: src/app/AuthChecker.tsx [hunk 19]
```diff
@@ -0,0 +1,32 @@
+import React, { useEffect } from 'react';
+
+import { useNavigate } from 'react-router-dom';
+
+import { useSelectUser } from '../features/auth/authSlice';
+
+import { useAppSelector } from './store';
+
+// import { auth } from '../integations/firebase';
```
Comment (remadyt) [inline src/app/AuthChecker.tsx]: why its needed?

File: src/app/AuthChecker.tsx [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  const user = useAppSelector(useSelectUser) ?? null;
+
+  useEffect(() => {
+    if (comp) {
+      if (user) return;
+      navigate(path);
+    }
+    if (!user) return;
+    navigate(path);
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
```
Comment (remadyt) [inline src/app/AuthChecker.tsx]: remove eslint errors

File: src/features/auth/authSlice.ts [hunk 21]
```diff
@@ -1,44 +1,56 @@
...
-        localStorage.setItem('user', JSON.stringify(user));
-
-        return user;
+      const data = await getFirestoreData<User>(DatabasePaths.USERS, 20);
+      const isExistUser = data.some((obj) => obj.nickName === nickName);
+
+      if (isExistUser) {
+        const existedUser = data.find((obj) => obj.nickName === nickName);
+
+        localStorage.setItem('user', JSON.stringify(existedUser));
```
Comment (remadyt) [inline src/features/auth/authSlice.ts]: create helper function for this

File: src/features/auth/authSlice.ts [hunk 22]
```diff
@@ -1,44 +1,56 @@
...
       }
-      notifications.show({
-        bg: 'cyan',
-        w: '450',
-        h: '80',
-        message: 'You must enter a unique nickName',
+
+      const collectionRef = collection(db, DatabasePaths.USERS);
+
+      await addDoc(collectionRef, {
```
Comment (remadyt) [inline src/features/auth/authSlice.ts]: why its needed?

File: src/features/components/Header/HeaderMegaMenu.module.css [hunk 23]
```diff
@@ -1,7 +1,7 @@
 .header {
-  height: rem(60px);
-  padding-left: var(--mantine-spacing-md);
-  padding-right: var(--mantine-spacing-md);
+  height: rem(110px);
+  padding-right: 50px;
```
Comment (remadyt) [inline src/features/components/Header/HeaderMegaMenu.module.css:3]: padding: 0 50px

File: src/features/components/Header/index.tsx [hunk 24]
```diff
@@ -19,22 +21,30 @@ import {
...
     useDisclosure(false);
   const [linksOpened, { toggle: toggleLinks }] = useDisclosure(false);
+  const [isUserShow, setIsUserShow] = useState(false);
+
   const theme = useMantineTheme();
 
   const { t } = useTranslation();
 
+  const toggleUserShow = () => {
+    setIsUserShow(!isUserShow);
```
Comment (remadyt) [inline src/features/components/Header/index.tsx]: prev => !prev

File: src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.module.css [hunk 25]
```diff
@@ -0,0 +1,17 @@
+.icon {
+	color: light-dark(var(--mantine-color-gray-5), var(--mantine-color-dark-3));
```
Comment (remadyt) [inline src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.module.css]: add stylelint

File: src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx [hunk 26]
```diff
@@ -0,0 +1,63 @@
+import { Avatar, Text, Group } from '@mantine/core';
+import { IconPhoneCall, IconAt, IconLogout } from '@tabler/icons-react';
+import { useDispatch } from 'react-redux';
+import { useNavigate } from 'react-router-dom';
+
+import { AppDispatch } from '../../../../../../app/store';
+import { getCurrentUser } from '../../../../../../shared/helpers/getCurrentUser';
```
Comment (remadyt) [inline src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx]: use global paths

File: src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+export const UserInfoIcons = ({ toggleUserShow }: Props) => {
+  const user = getCurrentUser();
+  const dispach: AppDispatch = useDispatch();
+  const navigate = useNavigate();
+
+  const handlelogOut = () => {
+    dispach(logOut());
+    navigate('/auth');
```
Comment (remadyt) [inline src/features/components/Header/ui/components/UserInfoIcons/UserInfoIcons.tsx]: if request error?

File: src/features/user/userSlice.ts [hunk 28]
```diff
@@ -1,26 +1,90 @@
...
+
+export const editUser = createAsyncThunk(
+  'user/editUser',
+  async (
+    { id, updatedUser }: { id: string; updatedUser: Partial<UserData> },
+    { rejectWithValue },
+  ) => {
+    try {
+      if (!id) {
+        throw new Error('Відсутній ідентифікатор користувача.');
```
Comment (remadyt) [inline src/features/user/userSlice.ts]: its bad ux, add notification

File: src/features/user/userSlice.ts [hunk 29]
```diff
@@ -1,26 +1,90 @@
...
+);
+
+export const removeUserById = createAsyncThunk(
+  'user/removeUserById',
+  async (userId: string, { rejectWithValue }) => {
+    const docRef: DocumentReference = doc(db, 'users', userId);
+
+    try {
+      await deleteDoc(docRef);
+      const data = await getFirestoreData<UserData>(DatabasePaths.USERS, 20);
```
Comment (remadyt) [inline src/features/user/userSlice.ts]: u take only 20? and implement deleting user not needed

File: src/pages/Auth/ui/components/AuthenticationTitle/Auth.tsx [hunk 30]
```diff
@@ -116,14 +103,7 @@ export const AuthenticationTitle = () => {
       case 'EnterOTP':
```
Comment (remadyt) [inline src/pages/Auth/ui/components/AuthenticationTitle/Auth.tsx]: use enum for cases

File: src/pages/Auth/ui/components/AuthenticationTitle/components/EnterNickName/EnterNickName.tsx [hunk 31]
```diff
@@ -1,53 +1,56 @@
...
-        className={classes.nickNameButton}
-      >
-        <IconArrowLeft
-          width={rem(50)}
-          height={rem(20)}
-          size={rem(10)}
-          stroke={rem(2)}
-          color="currentColor"
+    <Box className={classes.nickNameWrapper} maw={340} mx="auto">
+      <form onSubmit={form.onSubmit((value) => handleFormSubmit(value))}>
```
Comment (remadyt) [inline src/pages/Auth/ui/components/AuthenticationTitle/components/EnterNickName/EnterNickName.tsx]: form.onSubmit(handleFormSubmit)

File: src/pages/Auth/ui/components/AuthenticationTitle/components/EnterNickName/EnterNickName.tsx [hunk 32]
```diff
@@ -1,53 +1,56 @@
...
   const { t } = useTranslation();
 
+  const handleFormSubmit = ({
+    nickName,
+  }: {
+    nickName: string;
+    termsOfService: boolean;
+  }) => {
+    handleSubmit(nickName);
+    form.setValues({ nickName: '' });
```
Comment (remadyt) [inline src/pages/Auth/ui/components/AuthenticationTitle/components/EnterNickName/EnterNickName.tsx]: why its needed?

File: src/shared/helpers/editUser.ts [hunk 33]
```diff
@@ -0,0 +1,15 @@
...
+import { UserData } from '../types/Types';
+
+export const editUser = async (userId: string, editedUser: UserData) => {
+  try {
+    const docRef = doc(db, 'users', userId);
+
+    await updateDoc(docRef, editedUser);
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(`error ${error}`);
```
Comment (remadyt) [inline src/shared/helpers/editUser.ts]: add notification instead of console log

File: src/shared/helpers/editUser.ts [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+import { db } from '../../integations/firebase';
+import { UserData } from '../types/Types';
+
+export const editUser = async (userId: string, editedUser: UserData) => {
+  try {
+    const docRef = doc(db, 'users', userId);
+
+    await updateDoc(docRef, editedUser);
+  } catch (error) {
+    // eslint-disable-next-line no-console
```
Comment (remadyt) [inline src/shared/helpers/editUser.ts]: remove eslint error here and all such places

**Review by remadyt — CHANGES_REQUESTED** [review]

**Pull request merged**